pkg/driver/node_test.go

```diff
@@ -545,11 +553,66 @@ var _ = Describe("Node", func() {
 		})
 	})
 
-	It("should return unimplemented error for NodeGetVolumeStats", func() {
-		res, err := drv.NodeGetVolumeStats(ctx, &csi.NodeGetVolumeStatsRequest{})
-		Expect(res).To(BeNil())
-		status, ok := status.FromError(err)
-		Expect(ok).To(BeTrue())
-		Expect(status.Code()).To(Equal(codes.Unimplemented))
+	Describe("NodeGetVolumeStats", func() {
+		var req *csi.NodeGetVolumeStatsRequest
+
+		BeforeEach(func() {
+			req = &csi.NodeGetVolumeStatsRequest{
+				VolumeId:   volumeId,
+				VolumePath: "/volume/path",
+			}
+		})
+
+		It("should return an error if the volume ID is empty", func() {
+			req.VolumeId = ""
+			_, err := drv.NodeGetVolumeStats(ctx, req)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("Volume Id not provided"))
+			statusErr, ok := status.FromError(err)
+			Expect(ok).To(BeTrue())
+			Expect(statusErr.Code()).To(Equal(codes.InvalidArgument))
+		})
+
+		It("should fail if check for volumePath exists fails", func() {
+			mockOS.EXPECT().Exists(gomock.Any(), gomock.Any()).Return(false, errors.New("error"))
+			_, err := drv.NodeGetVolumeStats(ctx, req)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("failed to check whether volumePath exists"))
+			statusErr, ok := status.FromError(err)
+			Expect(ok).To(BeTrue())
+			Expect(statusErr.Code()).To(Equal(codes.Internal))
+		})
+
+		It("should fail if the volume path is empty", func() {
+			req.VolumePath = ""
+			_, err := drv.NodeGetVolumeStats(ctx, req)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("Volume path not provided"))
+			statusErr, ok := status.FromError(err)
+			Expect(ok).To(BeTrue())
+			Expect(statusErr.Code()).To(Equal(codes.InvalidArgument))
+		})
+
+		It("should fail if check GetDeviceStats fails", func() {
+			mockOS.EXPECT().Exists(gomock.Any(), gomock.Any()).Return(true, nil)
+			mockOS.EXPECT().Statfs(gomock.Any(), gomock.Any()).Return(errors.New("error"))
+			_, err := drv.NodeGetVolumeStats(ctx, req)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("failed to get stats by path"))
+			statusErr, ok := status.FromError(err)
+			Expect(ok).To(BeTrue())
+			Expect(statusErr.Code()).To(Equal(codes.Internal))
+		})
+
+		It("should return volume stats", func() {
+			mockOS.EXPECT().Exists(gomock.Any(), gomock.Any()).Return(true, nil)
+			mockOS.EXPECT().Statfs(gomock.Any(), gomock.Any()).Return(nil)
+			res, err := drv.NodeGetVolumeStats(ctx, req)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(res).To(SatisfyAll(
+				HaveField("Usage", Not(BeNil())),
```

Comment:
Can we somehow compare here the real usage values instead of just checking that something has been set?

Reply:
Updated test case